File: synth_ai/sdk/task/vendors.py
"""Vendor API key helpers shared by Task Apps."""

from __future__ import annotations

import os

from .errors import http_exception

_VENDOR_KEYS = {
    "OPENAI_API_KEY": ("dev_openai_api_key", "DEV_OPENAI_API_KEY"),
    "GROQ_API_KEY": ("dev_groq_api_key", "DEV_GROQ_API_KEY"),
}


def _mask(value: str, *, prefix: int = 4) -> str:
    if not value:
        return "<empty>"
    visible = value[:prefix]
    return f"{visible}{'…' if len(value) > prefix else ''}"


def _normalize_single(key: str) -> str | None:
    direct = os.getenv(key)
    if direct:
        return direct
    fallbacks = _VENDOR_KEYS.get(key, ())
    for env in fallbacks:
        candidate = os.getenv(env)
        if candidate:
            os.environ[key] = candidate
            print(
                f"[task:vendor] {key} set from {env} (prefix={_mask(candidate)})",
                flush=True,
            )
            return candidate
    return None


def normalize_vendor_keys() -> dict[str, str | None]:
    """Normalise known vendor keys from dev fallbacks and return the mapping."""

    resolved: dict[str, str | None] = {}
    for key in _VENDOR_KEYS:
        resolved[key] = _normalize_single(key)
    return resolved


def get_openai_key_or_503() -> str:
    key = _normalize_single("OPENAI_API_KEY")
    if not key:
        raise http_exception(503, "missing_openai_api_key", "OPENAI_API_KEY is not configured")
    return key


def get_groq_key_or_503() -> str:
    key = _normalize_single("GROQ_API_KEY")
    if not key:
        raise http_exception(503, "missing_groq_api_key", "GROQ_API_KEY is not configured")
    return key

File: synth_ai_core/src/localapi/vendors.rs
use std::collections::HashMap;
use std::env;

const VENDOR_KEYS: [(&str, [&str; 2]); 2] = [
    (
        "OPENAI_API_KEY",
        ["dev_openai_api_key", "DEV_OPENAI_API_KEY"],
    ),
    ("GROQ_API_KEY", ["dev_groq_api_key", "DEV_GROQ_API_KEY"]),
];

fn mask(value: &str, prefix: usize) -> String {
    if value.is_empty() {
        return "<empty>".to_string();
    }
    let visible: String = value.chars().take(prefix).collect();
    if value.len() > prefix {
        format!("{visible}…")
    } else {
        visible
    }
}

pub fn normalize_single(key: &str) -> Option<String> {
    if let Ok(value) = env::var(key) {
        if !value.is_empty() {
            return Some(value);
        }
    }
    let fallbacks = VENDOR_KEYS.iter().find(|(k, _)| *k == key);
    if let Some((_key, fallbacks)) = fallbacks {
        for env_key in fallbacks {
            if let Ok(value) = env::var(env_key) {
                if !value.is_empty() {
                    env::set_var(key, &value);
                    println!(
                        "[task:vendor] {} set from {} (prefix={})",
                        key,
                        env_key,
                        mask(&value, 4)
                    );
                    return Some(value);
                }
            }
        }
    }
    None
}

pub fn normalize_vendor_keys() -> HashMap<String, Option<String>> {
    let mut resolved = HashMap::new();
    for (key, _) in VENDOR_KEYS.iter() {
        resolved.insert((*key).to_string(), normalize_single(key));
    }
    resolved
}

pub fn get_openai_key() -> Option<String> {
    normalize_single("OPENAI_API_KEY")
}

pub fn get_groq_key() -> Option<String> {
    normalize_single("GROQ_API_KEY")
}
